qiime2/sdk/proxy.py

```diff
@@ -0,0 +1,193 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+from qiime2.core.type.util import is_visualization_type, is_collection_type
+
+
+class Proxy:
+    """Abstract base class to indicate that a given class that inherits from it
+    is a proxy
+    """
+    pass
+
+
+class ProxyResult(Proxy):
+    def __init__(self, future, selector, signature=None):
+        """We have a future that represents the results of some QIIME 2 action,
+        and we have a selector indicating specifically which result we want
+        """
+        self._future_ = future
+        self._selector_ = selector
+        self._signature_ = signature
+
+    @property
+    def _archiver(self):
+        return self.result()._archiver
+
+    @property
+    def type(self):
+        return self._archiver.type
+
+    @property
+    def uuid(self):
+        return self._archiver.uuid
+
+    @property
+    def format(self):
+        return self._archiver.format
+
+    @property
+    def citations(self):
+        return self._archiver.citations
+
+    def result(self):
+        return self.get_element(self._future_.result())
+
+    def get_element(self, results):
+        """Get the result we want off of the future we have
+        """
+        return getattr(results, self._selector_)
+
+    def save(self, filepath, ext=None):
+        """Blocks then calls save on the result.
+        """
+        return self.result().save(filepath, ext=ext)
+
+
+class ProxyArtifact(ProxyResult):
+    """This represents a future Artifact that is being returned by a Parsl app
+    """
+    def __repr__(self):
+        if self._signature_ is None:
+            return f'<artifact: Unknown Type {object.__repr__(self)}>'
+        else:
+            return \
+                f'<artifact: {self._signature_[self._selector_].qiime_type}>'
```

Comment:
Technically, we want the type that `signature.solve_outputs()` comes up with for this selector.

Reply:
I need a signature object to do that, but I don't have access to one. Should I just use the entire Signature object for _signature_ instead of just the outputs or something? That would make more sense with the name anyway lol.

Reply:
You would pass this in before it's evaluated by Parsl in theory. So you have the inputs, you can use solve and annotate the proxies with this new info

Reply:
So `_signature_` would be the solved outputs?

Reply:
yeah I think so